wikimedia/wikipedia-ios
# Better Block Message Support

**Phabricator:** 
https://phabricator.wikimedia.org/T169013
https://phabricator.wikimedia.org/T275118
https://phabricator.wikimedia.org/T201640

### Notes
This PR adds improved display of blocked messages to users when creating an account, publishing an article description, and publishing an edit in our section editor. It adjusts the API calls in these areas and parses the error responses. It also adds a new blocked message panel (from our existing Panels.swift set of views) and disables the editors when the user is blocked.

This PR is best reviewed commit-by-commit. I tried to leave detail in the commit messages. Any developer testing should be tested with all commits in place. 

There are a couple of things remaining after this PR for blocked messages:

1. Showing a confirmation if the user is going to lose their edits after tapping a link in the panel, which will be integrated after #4454 is merged.
2. We have some url routing issues with some of the blocked messages, like our native talk page intercepting a link like https://en.wikipedia.org/wiki/Special:MyTalk. I put up a couple of temporary fixes for EN, but a better approach needs to be in place for other language support (see Phab [task](https://phabricator.wikimedia.org/T328678)).

### Test Steps
1. Block one of your accounts on [Test Wikipedia](https://test.wikipedia.org/wiki/Special:Block) (if you don't have admin privileges, I can block one for you) and on [Test Wikidata](https://test.wikidata.org/wiki/Special:Block).
2. Log into that account on device. Set up your primary app language as non-Test language (this is so that the correct MediaWiki:Blockedtext template is fetched after blocked API error. Test Wikipedia doesn't have one set up).
4. Visit an article on Test Wikipedia. Go to section editor. Confirm blocked panel appears after landing on editor, and editor is disabled.
5. Go to article description editor by tapping the first pencil (you may need to hunt around for a Test article with a description already in place...there's something on the PCS-side that doesn't show the "Add article description" button for those without).
6. Confirm blocked panel appears after landing on article description editor, and description editing is disabled.
7. Temporary adjust code so that article description editor updates on Wikidata rather than on Test Wikipedia. The changes in [this temp branch commit](https://github.com/wikimedia/wikipedia-ios/commit/ac6850f7d2912e8fe9d199659b6c30647427e51c) will do it.
8. Repeat steps 4 & 5. Confirm blocked panel still appears.
9. Brief regression testing - unblock account on Test Wikipedia and confirm you can successfully publish an edit in section editor.
10. Now test create account blocked error. Set up app languages to only contain Test Wikipedia. Then re-block account, making sure there's also an IP block in place.
11. Try to create an account from our create account screen. Confirm blocked message appears after attempting to create an account.

_Note: I also have blocked panels showing upon Publish, not just landing on the editors. Test steps for this will be a bit tedious - basically repeating 1-8, but unblocking your account in between each step so you can get past disabling of the editors. I didn't include these steps in the interest of leaning on QA more for this._

### Screenshots

Create Account Error
![IMG_0157](https://user-images.githubusercontent.com/3620196/216197336-a1df97ed-aee3-4186-a30b-0043b2dc2d7a.PNG)

Section Editor Error
![IMG_0156](https://user-images.githubusercontent.com/3620196/216197342-a17266e0-3181-470f-9733-9ff3933e8746.PNG)
